### DISCUSSION TITLE
Get object from output

### DISCUSSION THREAD

**Discussion by ag4ve** [xonsh/xonsh#5531] [Q&A] [answered]
I defined this alias:
```xsh
@aliases.register
def _yamlstr(args, stdin=None, stdout=None):
    if stdin is None:
        return
    data = yaml.safe_load(stdin.read())

    if len(args) < 1:
        return data
    else:
        globals()[args[0]] = data
```

Which obviously works fine if I just pass it a variable to use. But I can't figure out how to just define a variable with this:
`s = dig +yaml google.com | yamlstr`
(I've tried parenthesis and $(...) etc - generally xonsh crashes)

So, is this type of workflow possible or do I just pass in a variable name and define globals()[name]?

**Comment by anki-code** [comment, marked as answer]
In xonsh 0.17.0 there is the way to format command output and I'm going to give you more broader example from [xonsh-cheatsheet](https://github.com/anki-code/xonsh-cheatsheet/blob/main/README.md):

```xsh
from xonsh.procs.specs import SpecAttrModifierAlias
imp = type('ImpCl', (object,), {'__getattr__':lambda self, name: __import__(name) })()  # Sugar for inline import

aliases['@lines'] = SpecAttrModifierAlias({"output_format": 'list_lines'},
                                           "Set `list_lines` output format.")
aliases['@json'] = SpecAttrModifierAlias({"output_format": lambda lines: imp.json.loads('\n'.join(lines))},
                                           "Set `json` output format.")
aliases['@path'] = SpecAttrModifierAlias({"output_format": lambda lines: imp.pathlib.Path(':'.join(lines))},
                                           "Set `path` output format.")
aliases['@yaml'] = SpecAttrModifierAlias({"output_format": lambda lines: imp.yaml.safe_load('\n'.join(lines))},
                                           "Set `yaml` output format.")
```
Now you can:
```xsh
y = $(@yaml dig +yaml google.com)
y[0]['message']['query_time']
# datetime(2024)

aliases['ydig'] = '@yaml dig +yaml'  # You can create final shortcut.

y = $(ydig google.com)
y[0]['type']
# 'MESSAGE'
```
and as well
```xsh
$(@lines ls /)
# ['/bin', '/etc', '/home']

$(@json echo '{"a":1}')  # Try with `curl` ;)
# dict({"a":1})

$(@path which xonsh)
# Path('/path/to/xonsh')

$(@path which xonsh).parent
# Path('/path/to')
```
Enjoy.

You can also just install [xontrib-spec-mod](https://github.com/anki-code/xontrib-spec-mod).

**Reply by anki-code** [reply]
If you need direct answer to your initial question of how to achieve this with callable alias. The short answer - callable alias can't return object and you need to convert the output manuallly e.g. `y = to_yaml($(dig))`. Also take a look into [xontrib-pipeliner](https://github.com/anki-code/xontrib-pipeliner).

**Reply by ag4ve** [reply]
Thanks, I think that means that I just have to return a SpecAttrModifierAlias when calling @yaml or @json etc and it should work or:
`s = $(@yaml dig +yaml google.com)` # if i process stdin like i was - maybe (sorry, I'm a bit tired and will have to pick this up tomorrow). I really appreciate the answer and module and cheatsheet update too.
